src/main/java/com/google/devtools/build/lib/syntax/SkylarkList.java

```diff
@@ -201,6 +206,312 @@ public int hashCode() {
+        private final int stop;
+        private final int step;
+
+        private Slice(int start, int stop, int step) {
+          this.start = start;
+          this.stop = stop;
+          this.step = step;
+        }
+
+        /**
+         * Computes slice indices for the requested range slice.<p>
+         *
+         * The implementation is based on CPython
+         * https://github.com/python/cpython/blob/09bb918a61031377d720f1a0fa1fe53c962791b6/Objects/sliceobject.c#L366-L509
+         **/
+        public static Slice from(int length, Object startObj, Object endObj, Object stepObj, Location loc)
+            throws EvalException {
+          int start;
+          int stop;
+          int step;
+
+          if (stepObj == Runtime.NONE) {
+            step = 1;
+          } else if (stepObj instanceof Integer) {
+            step = (Integer) stepObj;
+          } else {
+            throw new EvalException(
+                loc, String.format("slice step must be an integer, not '%s'", stepObj));
+          }
+          if (step == 0) {
+            throw new EvalException(loc, "slice step cannot be zero");
+          }
+
+          int upper;
+          int lower;
+          if (step < 0) {
+            lower = -1;
+            upper = length + lower;
+          } else {
+            lower = 0;
+            upper = length;
+          }
+
+          if (startObj == Runtime.NONE) {
+            start = step < 0 ? upper : lower;
+          } else if (startObj instanceof Integer) {
+            start = (Integer) startObj;
+            if (start < 0) {
+              start += length;
+              start = Math.max(start, lower);
+            } else {
+              start = Math.min(start, upper);
+            }
+          } else {
+            throw new EvalException(
+                loc, String.format("slice start must be an integer, not '%s'", startObj));
+          }
+          if (endObj == Runtime.NONE) {
+            stop = step < 0 ? lower : upper;
+          } else if (endObj instanceof Integer) {
+            stop = (Integer) endObj;
+            if (stop < 0) {
+              stop += length;
+              stop = Math.max(stop, lower);
+            } else {
+              stop = Math.min(stop, upper);
+            }
+          } else {
+            throw new EvalException(loc, String.format("slice end must be an integer, not '%s'", endObj));
+          }
+          return new Slice(start, stop, step);
+        }
+      }
+
+      private final int start;
+      private final int stop;
+      private final int step;
+      private final int size;
+
+      private RangeListView(int start, int stop, int step) {
+        this.start = start;
+        this.stop = stop;
+        this.step = step;
+        this.size = computeSize(start, stop, step);
+      }
+
+      @Override
+      public Integer get(int index) {
+        int value = start + step * index;
+        if ((step > 0 && value > stop) || (step < 0 && value < stop)) {
```

Comment:
If stop is exclusive, shouldn't it be if `value >= stop` and `value <= stop`?

Then again, why do the computation on values at all? Why not just test if `index < 0 || index >= size`?

Reply:
great point! I didn't realize that index that is passed here is already normalized